### ISSUE TITLE
dead_code lint incorrectly warns about struct field with side effects in drop-impl

### ISSUE THREAD

**Issue by LukasKalbertodt** [rust-lang/rust#39129]
I tried this code ([playground](https://play.rust-lang.org/?gist=29e5b9f3d26f2b8f11b9d8386ff7d7ee&version=stable&backtrace=0)):

```rust
struct HappyDrop;

impl Drop for HappyDrop {
    fn drop(&mut self) {
        println!("❤ I have been dropped! ╰( ◕ ᗜ ◕ )╯");
    }
}

struct Foo {
    drop: HappyDrop,
    stuff: u8,
}

fn main() {
    let x = Foo { 
        drop: HappyDrop,
        stuff: 8,
    };
    println!("{}", x.stuff);
}
```

I expected to see this happen: No warning, because every struct field serves a purpose.

Instead, this happened: 

```
warning: field is never used: `drop`, #[warn(dead_code)] on by default
  --> <anon>:10:5
   |
10 |     drop: HappyDrop,
   |     ^^^^^^^^^^^^^^^
```

My example is pretty useless, but some side effects in `drop()` might be really useful.

## Meta

Tested on `rustc 1.14.0 (e8a012324 2016-12-16)` and `1.16.0-nightly (ff591b6dc 2017-01-15)`.

Backtrace:

**Comment by KalitaAlexey** [comment]
@LukasKalbertodt,
It is expected.
You can write `#[allow(dead_code)]` to suppress that.

**Comment by LukasKalbertodt** [comment]
Thanks, but I know, and I guess `#[allow()]` attributes shouldn't be overused. Is it really expected? It looks like a confusing false positive to me 🤔

**Comment by KalitaAlexey** [comment]
@LukasKalbertodt,
`String` has `Drop`.
Shouldn't the compiler warn you if you don't use it?

**Comment by LukasKalbertodt** [comment]
@KalitaAlexey Good point. I guess it's not really possible to improve this lint. My bad, I haven't thought this through, sorry!
